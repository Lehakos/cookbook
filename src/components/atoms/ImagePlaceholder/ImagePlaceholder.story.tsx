import React from 'react';
import { storiesOf } from '@storybook/react';

import ImagePlaceholder from './ImagePlaceholder';

const stories = storiesOf('Atoms|ImagePlaceholder', module);

stories.add('Default', () => <ImagePlaceholder />);
