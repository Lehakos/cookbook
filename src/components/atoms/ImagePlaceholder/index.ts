export { default, Props } from './ImagePlaceholder';
