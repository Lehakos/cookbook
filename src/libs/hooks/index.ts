export { useControlledProp } from './use-controlled-prop';
export { useElementSize } from './use-element-size';
