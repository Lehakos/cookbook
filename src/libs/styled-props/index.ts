export { typography, TypographyProps } from './typography';
export { layout, LayoutProps } from './layout';
export { flexbox, FlexboxProps } from './flexbox';
