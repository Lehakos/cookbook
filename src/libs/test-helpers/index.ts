export { render } from './render';
export { mockElementWithBoundingRect } from './mock-element-with-bounding-rect';
