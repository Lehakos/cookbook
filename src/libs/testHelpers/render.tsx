import React, { ReactElement } from 'react';
import { render as baseRender } from '@testing-library/react';
import { ThemeProvider } from 'styled-components';

import theme from 'shared/theme';

export const render = (el: ReactElement) =>
  baseRender(<ThemeProvider theme={theme}>{el}</ThemeProvider>);
