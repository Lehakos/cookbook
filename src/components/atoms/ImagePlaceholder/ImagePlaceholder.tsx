import React, { Children } from 'react';

import { StrictUnion } from 'shared/types';

import * as s from './ImagePlaceholder.styles';

interface Size {
  width: number;
  height: number;
}

export type Props = {
  responsive?: boolean;
} & StrictUnion<Size | {}>;

export const DEFAULT_WIDTH = 480;
export const DEFAULT_HEIGHT = 360;

const ImagePlaceholder: React.FunctionComponent<Props> = function({
  width,
  height,
  responsive,
  children,
}) {
  const computedWidth = width || DEFAULT_WIDTH;
  const computedHeight = height || DEFAULT_HEIGHT;
  const hasChildren = Children.toArray(children).some(Boolean);

  const getWidth = () => {
    if (responsive) return '100%';
    return `${computedWidth}px`;
  };

  const getHeight = () => {
    if (!responsive || !hasChildren) return `${computedHeight}px`;
    return 'auto';
  };

  return (
    <s.Wrapper width={getWidth()} height={getHeight()}>
      {hasChildren ? (
        children
      ) : (
        <s.Size aria-hidden>
          {computedWidth} x {computedHeight}
        </s.Size>
      )}
    </s.Wrapper>
  );
};

export default ImagePlaceholder;
