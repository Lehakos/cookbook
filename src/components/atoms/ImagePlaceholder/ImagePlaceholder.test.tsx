import React from 'react';

import { render } from 'libs/testHelpers';

import ImagePlaceholder, {
  Props,
  DEFAULT_HEIGHT,
  DEFAULT_WIDTH,
} from './ImagePlaceholder';

const WIDTH = 640;
const HEIGHT = 480;

const renderComp = (props: Partial<Props> = {}) => {
  const defaultProps: Props = {
    width: WIDTH,
    height: HEIGHT,
  };

  return render(<ImagePlaceholder {...defaultProps} {...props} />);
};

describe('<ImagePlaceholder />', () => {
  it('правильно отображается', () => {
    const { container } = render(<ImagePlaceholder />);
    expect(container.firstChild).toMatchSnapshot();
  });

  describe('если переданы пропсы width и height', () => {
    it('показывает заглушку с этими размерами', () => {
      const { queryByText } = renderComp();
      expect(queryByText(`${WIDTH} x ${HEIGHT}`)).toBeTruthy();
    });
  });

  describe('если не переданы пропсы width и height', () => {
    it('показывает заглушку с дефолтными размерами', () => {
      const { queryByText } = renderComp({
        width: undefined,
        height: undefined,
      });
      expect(queryByText(`${DEFAULT_WIDTH} x ${DEFAULT_HEIGHT}`)).toBeTruthy();
    });
  });
});
