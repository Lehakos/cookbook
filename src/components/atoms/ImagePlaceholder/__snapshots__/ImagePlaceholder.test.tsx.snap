// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ImagePlaceholder /> правильно отображается 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-content: center;
  -ms-flex-line-pack: center;
  align-content: center;
  width: 480px;
  height: 360px;
  text-align: center;
  background-color: #CFD8D7;
}

.c1 {
  text-transform: uppercase;
  -webkit-letter-spacing: 0.1em;
  -moz-letter-spacing: 0.1em;
  -ms-letter-spacing: 0.1em;
  letter-spacing: 0.1em;
}

<div
  class="c0"
  height="360px"
  width="480px"
>
  <span
    aria-hidden="true"
    class="c1"
  >
    480
     x 
    360
  </span>
</div>
`;
